### DISCUSSION TITLE
FileUpload in S3 does not work, and no error is rised

### DISCUSSION THREAD

**Discussion by philomath27** [filamentphp/filament#10380] [Help]
### Package

filament/filament

### Package Version

3.0

### Laravel Version

10.10

### PHP Version

8.1

### Problem description

I just configured my app to upload images to Amazon s3 (Simple Storage Service), with the Facade in laravel Storage it works, but in my app is not working, and no error is raised, it's extremely rare and i don't know how to fix it.

### Expected behavior

Image is uploaded to S3 and showed in the Upload component (image)

### Steps to reproduce

Just a form with a Image and a bucket repository in Amazon S3 

### Reproduction repository

https://github.com/philomath27/escolare

**Comment by danharrin** [comment]
It is very likely that it is a misconfiguration of your bucket and not something a reproduction repository will help with. Many users use S3 with Filament successfully in production. I think I'm going to convert this into a discussion, and users who wish can help you debug what is wrong. Please @ me if you find out that this objectively is not working for any S3 users.

**Reply by leandrocfe** [reply]
@philomath27 share the repo please.

**Reply by brentkelly** [reply]
@danharrin FYI while you are correct - in that it is a s3 misconfiguration issue - it is also the **default recommended configuration** of an S3 bucket (as directed when setting up S3). Following the S3 guidance when setting up your bucket results in a situation that Livewire (and thus Filament) considers "misconfigured".

Worse still, the default S3 guidance when setting up a new bucket results in:
1. Filament uploads *appearing* to work - yet files not arriving in your bucket.
2. No errors in the error log.

This is seriously confusing for the average punter. Unless you set `filesystems.disks.s3.throw` to `true` in your Laravel configuration you will get no errors, and nothing in your error log. If you turn it on, the error log will finally then report "The bucket does not allow ACLs" - and you can make some headway.

Because this is default Livewire behavior (Filament is not specifying any visibility state by default so Livewire defaults to setting ACL to 'public'), I'm not sure there is much Filament can do - aside from perhaps flagging it in the documentation?

I've tried to at least update the various discussions / posts on this so if people Google it they might run into the solution / workaround.

See my response at the bottom of this thread, and here https://github.com/filamentphp/filament/discussions/11486#discussioncomment-13568916 for more detail as to the issue & the workaround I use to avoid the problem.

**Comment by jgmuchiri** [comment]
For me, this was caused by S3 configuration. It was uploading to the temp directory but not uploading to bucket after saving.
- Unchecked "Block all public access"
- Set bucket policy
- Setup  CORS
- Enable ACLs -> Bucket owner preferred (This seems to be the culprit)

After these changes, filament is uploading. Note the files are public **publicly** accessible.

**Reply by jgmuchiri** [reply]
## Policy

```json
{
    "Version": "2012-10-17",
    "Id": "Policy1716477330832",
    "Statement": [
        {
            "Sid": "Stmt1716477329263",
            "Effect": "Allow",
            "Principal": "*",
            "Action": "s3:*",
            "Resource": [
                "arn:aws:s3:::products",
                "arn:aws:s3:::products/*"
            ]
        }
    ]
}
```


## Cors
```json
[
    {
        "AllowedHeaders": [
            "*"
        ],
        "AllowedMethods": [
            "GET",
            "HEAD",
            "PUT",
            "POST",
            "DELETE"
        ],
        "AllowedOrigins": [
            "*"
        ],
        "ExposeHeaders": [],
        "MaxAgeSeconds": 3000
    }
]
```

**Reply by jgmuchiri** [reply]
<img width="822" alt="image" src="https://github.com/filamentphp/filament/assets/4142133/00c167d8-82e8-4118-b03c-3c8ab4007b47">

**Reply by Jartz** [reply]
it worked, thank you

**Comment by brentkelly** [comment]
FYI the issue here is that by default Livewire will attempt to use S3 ACL to set public visibility when you upload a file. However this throws an error unless your S3 bucket is configured to use ACLs - **something that S3 recommends against as default**.

This issue is compounded by Laravel's default filesystem configuration for s3 to set `throw` to `false` - meaning the resulting error is suppressed.

What you end up with is file uploads in Filament appearing to work - but not working, with no errors in the error log. Seriously confusing to the average punter.

There are two solutions:
1. Enable ACL on your bucket as discussed above; or
2. Set `visibility` on your `FileUpload` component to any string other than `public` or `private` (empty string will work - but `null` won't). There are two ways of doing this - either calling the `visibility` method, or creating a custom `FileUpload` component that defaults the `visibility`.

Example of one-off / inline use:
```php
FileUpload::make('foo')
    ->visibility('oh-please-let-this-work')
    // etc
```

Most use cases I've had I want to upload images & files that are then publicly available - so I follow the S3 recommendation of disabling ACL & using a `s3:GetObject` bucket policy to grant public access to all files in my bucket. Because of this I typically want all FileUpload components in my project to have a visbility of neither public nor private - so I tend to create a custom class. See my answer in this discussion for an example of that: https://github.com/filamentphp/filament/discussions/11486#discussioncomment-13568916
